README.md

```diff
@@ -111,6 +111,129 @@ The custom middleware can then be configured like the following:
 Rails.application.config.app_profiler.middleware = AppProfilerAuthorizedMiddleware
 ```
 
+## Profile Server
+
+This option allows for profiles to be passively collected via an HTTP endpoint,
+inspired by [golang's built-in pprof server](https://pkg.go.dev/net/http/pprof).
+
+A minimal Rack app runs a minimal (non-compliant) HTTP server, which exposes an
+endpoint that allows for profiling. For security purposes, the server is bound
+to localhost only. The HTTP server is built using standard library modules only,
+in order to keep dependencies minimal. Because it is an entirely separate server,
+listening on an entirely separate socket, this should not interfere with any
+existing application routes, and should even be usable in non-web apps.
+
+This allows for two main use cases:
+
+- Passive profile collection in production
+  - Periodically profiling production apps to analyze them responding to real
+workloads
+  - Providing a statistical, long-term view of the "hot paths" of the workload
```

Comment:
"Real workloads" refers to that this profiles the actual requests an application is receiving organically, ie, not a specific endpoint that was asked to be profiled, which is somewhat a "synthetic" request.

"Hot paths" indicates what parts of the code are showing up in the most stack traces, ie, regardless of what request is being performed. This might indicate a specific request is slow, or that a shared path used by different related requests is slow, etc.

Does that help to clarify?

Reply:
Yes, but I think it would help me to see what kind of profile data this produces in comparison to the traditional approach of profiling a request of a Rails app. Since the thread just sleeps, I assumed it would just be a blank profile.

Reply:
> what kind of profile data this produces in comparison to the traditional approach of profiling a request of a Rails app

TL;DR, it will provide for aggregated / averaged profiles (see the examples at the end), displayed in a format similar to the "left heavy" view in speedscope. It will show an aggregate view of **everything** the app is doing during these profiling sessions.

>  Since the thread just sleeps, I assumed it would just be a blank profile.

The sleeping thread shouldn't show up much, it should predominantly show threads that are actually doing work, ie, 
 this will profile the **other threads** in the application, though more generally, it will get the stack trace for "whatever is running at the time". My understanding of stackprof is that:

- For CPU and Wall modes, an `itimer` is set (based on the configured "interval") then it will get whatever stack is the "active" one in the ruby VM at the time (stackprof itself works via interrupts / tracepoints, so doesn't have its own "thread", which is why it itself never shows up)
  - For Wall mode, the timer is based on real time elapsing
  - For CPU mode, the timer should only be "charged" when a thread is actually running
- For Object mode, it will be based on a certain number of allocations (identified by the sample rate) Eg, "1" would be for every object allocation, for "1000", it would be every thousandth allocation, etc.

Note that it is still possible to see "sleep" in the CPU profile, because a sleeping thread in ruby will [periodically awaken, then return to a native sleep](https://github.com/ruby/ruby/blob/master/thread.c#L1264), and this might coincide with the timer expiring given enough samples.

For Object mode, it is based on tracepoints and not a timer, so we will only see the stacks for other threads which are actually allocating objects (we won't see the profile server, since it is sleeping and therefore not allocating anything).

I think the most interesting data here is the CPU and object allocation, which should tell you:

- CPU: Where is your application actually spending its CPU time when it gets a chance to run?
- Object: What code paths in your application are responsible for allocating objects (how many?)

Here are some samples of real data collected (albeit, against a relatively idle test app) with this, displayed in cloud profiler. The samples were collected approximately once a minute, for approximately 10 seconds. These should be the equivalent of the "left heavy" view in speedscope.

CPU graph (basically showing what an idle server uses its CPU time for, not terribly exciting, would be more interesting if we had some requests happening, but this is a good "baseline"):

<img width="1457" alt="Screen Shot 2022-06-14 at 2 00 15 PM" src="https://user-images.githubusercontent.com/618615/173658248-c673e619-f64d-4698-989c-1efb1fcad184.png">


Object profile graph:

<img width="1457" alt="Screen Shot 2022-06-14 at 2 01 45 PM" src="https://user-images.githubusercontent.com/618615/173658259-a8bfb2a8-32ec-4687-a4ef-fbb3d2f026cb.png">

The object profile graph shows the count of objects allocated on average by each stack frame, eg, the selected frame in the UI here had ~ 4 objects on average.

Note that both of these graphs are the aggregate of many profiling sessions, each representing the average of ~ 250 sessions. This thus gives us a pretty good average view of what threads are actually doing work in the application, and where allocations are happening. The larger the number of samples, the better the view.

Again, of course, the data would be more interesting for an active app, as we are seeing the CPU graph dominated by threads periodically waking up from their sleep (including our own profile server), and a relatively small number of object allocations.

Does this help to clarify a bit? This is a relatively novel use case for ruby profiling both at Shopify and I think in general, but exists for other languages (eg, golang as I have mentioned), so I believe it will be useful for developers once we expose it, as this sort of data has been used for Go apps for a few years now.